### PR TITLE
feat: support overriding service.name and service.version using command line flags

### LINKED ISSUE #8: Support overriding the OTEL service name
As @v1v pointed out, we could support setting a different service name, instead of the default `junit2otlp`:

1. using a CLI flag: https://github.com/mdelapenya/junit2otlp/pull/7#discussion_r750517623
2. using the `OTEL_SERVICE_NAME` environment variable https://github.com/mdelapenya/junit2otlp/pull/7#discussion_r750519506

Let's define the precedence order in the case both are defined

### PR DESCRIPTION
**Pull request by mdelapenya**
## What is this PR doing?
It adds two command-line flags: `--service-name`, `--service-version` and `--trace-name`, so that it's possible to define the service name and version when sending metrics, traces and spans.

It will read from user args, and if the specific `OTEL_SERVICE_`+suffix (NAME, VERSION)  variable is **NOT** present in the environment, then the value of the flag will be used as service name/version. We are adding unit tests for that evaluation.

These service name and version, read once, will be used to create both the traces provider and the metrics pusher as part of an SDK Resource entity, which will be passed to the initialiser methods.

To run the unit tests we are including, a Github action is added to run on each merge and PR, running the unit tests using `gotestsum`.

Finally, we are adding an integration test for the collection of traces. This test uses Testcontainers to spin up otel-exporter and Jaeger, defining a file exporter receiving the traces as JSON objects. The test will read that JSON file and will perform several asserts against it, like number of spans, or certain attributes obtained from the golden file used as sample (`TEST-sample.xml`).

To support reading that file in tests, and passing it to the execution, we have created a Reader interface: in runtime it will use a PipeReader, reading from a pipe, and in test-time it will directly read from a static file in the workspace.

## Related issues
- Closes #8

### CODE REVIEW

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+        uses: actions/setup-go@v2
+        with:
+          go-version: 1.16
+        id: go
+
+      - name: Get dependencies
+        run: |
+          go get -u gotest.tools/gotestsum
+          go get google.golang.org/grpc/internal/transport@v1.41.0
+          go get google.golang.org/grpc@v1.41.0
```
Comment (mdelapenya) [inline .github/workflows/tests.yml:29]: I do not like this, but this is the only way I've found to discover those modules: I added them locally, but after running `go mod tidy` they are pruned from the modules, causing the Go process to not discover them. 🤷

File: main.go [hunk 2]
```diff
@@ -109,6 +117,26 @@ func createTracesAndSpans(ctx context.Context, tracesProvides *sdktrace.TracerPr
 	return nil
 }
 
+// getOtlpEnvVar checks if the env variable, removing the OTEL prefix, needs to override
+func getOtlpEnvVar(key string, fallback string) string {
+	envVar := os.Getenv("OTEL_" + key)
```
Comment (v1v) [inline main.go:122]:
```suggestion
	envVar := os.Getenv(key)
```

File: main.go [hunk 3]
```diff
@@ -109,6 +117,26 @@ func createTracesAndSpans(ctx context.Context, tracesProvides *sdktrace.TracerPr
...
+	if envVar != "" {
+		return envVar
+	}
+
+	return fallback
+}
+
+// getOtlpServiceName checks the service name
+func getOtlpServiceName() string {
+	return getOtlpEnvVar("SERVICE_NAME", serviceNameFlag)
```
Comment (v1v) [inline main.go:132]:
```suggestion
	return getOtlpEnvVar("OTEL_SERVICE_NAME", serviceNameFlag)
```

File: main.go [hunk 4]
```diff
@@ -109,6 +117,26 @@ func createTracesAndSpans(ctx context.Context, tracesProvides *sdktrace.TracerPr
...
+}
+
+// getOtlpServiceVersion checks the service version
+func getOtlpServiceVersion() string {
+	return getOtlpEnvVar("SERVICE_VERSION", serviceVersionFlag)
```
Comment (v1v) [inline main.go:137]:
```suggestion
	return getOtlpEnvVar("OTEL_SERVICE_VERSION", serviceVersionFlag)
```

File: main.go [hunk 5]
```diff
@@ -43,9 +51,9 @@ func createIntCounter(meter metric.Meter, name string, description string) metri
 		)
 }
 
-func createTracesAndSpans(ctx context.Context, tracesProvides *sdktrace.TracerProvider, suites []junit.Suite) error {
-	tracer := tracesProvides.Tracer("junit2otlp")
-	meter := global.Meter("junit2otlp")
+func createTracesAndSpans(ctx context.Context, srvName string, tracesProvides *sdktrace.TracerProvider, suites []junit.Suite) error {
+	tracer := tracesProvides.Tracer(srvName)
```
Comment (v1v) [inline main.go:55]: I wonder if the traceName should have a different attribute:
- Service Name where all the traces will be sent to
- Trace Name, could be the same but customisable to a different one.
What do you think?

Comment (mdelapenya) [reply main.go:55]: I like the idea of overriding the trace name, will add it here in the PR, thanks

Comment (mdelapenya) [reply main.go:57]: Added in here: https://github.com/mdelapenya/junit2otlp/pull/9/commits/0428eb487ee955a54aadf3da4085c718dc274f94

I was not able to find an env var to set the trace name, i.e. `OTEL_TRACE_NAME`, that's why I'm directly using what's coming from the flag (`junit2otlp` as fallback) Please let me know if you know how to configure that trace name using the env

File: main.go [hunk 6]
```diff
@@ -156,7 +189,13 @@ func propsToLabels(props map[string]string) []attribute.KeyValue {
 	return attributes
 }
 
-func readFromPipe() ([]byte, error) {
+type InputReader interface {
```
Comment (mdelapenya) [inline main.go:192]: Adding an interface to support mocking this in tests

File: main_test.go [hunk 7]
```diff
@@ -0,0 +1,316 @@
...
+				Kind              string               `json:"kind"`
+				StartTimeUnixNano string               `json:"startTimeUnixNano"`
+				EndTimeUnixNano   string               `json:"endTimeUnixNano"`
+				Attributes        []TestTraceAttribute `json:"attributes"`
+				Status            struct {
+				} `json:"status"`
+			} `json:"spans"`
+		} `json:"instrumentationLibrarySpans"`
+	} `json:"resourceSpans"`
+}
```
Comment (mdelapenya) [inline main_test.go:74]: Struct objects mapped from the exporter. I understand that this struct could change if the otel-exporter is changed

File: main_test.go [hunk 8]
```diff
@@ -0,0 +1,316 @@
...
+	assert.Equal(t, "SPAN_KIND_INTERNAL", aTestCase.Kind)
+
+	codeFunction, _ := findAttributeInArray(aTestCase.Attributes, "code.function")
+	assertStringValueInAttribute(t, codeFunction.Value, "TestCheckConfigDirsCreatesWorkspaceAtHome")
+
+	testClassName, _ := findAttributeInArray(aTestCase.Attributes, "test.classname")
+	assertStringValueInAttribute(t, testClassName.Value, "github.com/elastic/e2e-testing/cli/config")
+
+	goVersion, _ := findAttributeInArray(aTestCase.Attributes, "go.version")
+	assertStringValueInAttribute(t, goVersion.Value, "go1.16.3 linux/amd64")
```
Comment (mdelapenya) [inline main_test.go:238]: Discarding the returned error, as the next assertion would fail

**Review by kuisathaverat — APPROVED** [review]

**Comment by mdelapenya** [comment]
@v1v I'm merging this one to progress and resolve possible conflicts in already opened PRs. Please open an issue if you have any other concern. Thanks!

**Pull request merged**